### PR TITLE
[pigeon] Minor fixes to NNBD type handling in Dart

### PR DESCRIPTION
**Pull request by stuartmorgan**
Fixes invalid Dart output for the following cases:
- Non-nullable enum types as fields
- Non-nullable custom class types as fields
- Nullable collection types as return values

Fixes https://github.com/flutter/flutter/issues/101515
Fixes https://github.com/flutter/flutter/issues/100594

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
(This deliberately skips from 3.0.0 to 3.0.2 to make it easier to resolve with https://github.com/flutter/packages/pull/1532, which I will land first.)

File: packages/pigeon/pigeons/nullable_returns.dart [hunk 1]
```diff
@@ -26,3 +26,23 @@ abstract class NullableArgHostApi {
 abstract class NullableArgFlutterApi {
   int doit(int? x);
 }
+
+@HostApi()
+abstract class NonNullCollectionHostApi {
+  List<String?>? doit();
+}
```
Comment (gaaclarke) [inline packages/pigeon/pigeons/nullable_returns.dart:33]: Adding this here is making sure that the generated code compiles and the dart unit tests below make sure it is generating code as you expect it.  But we aren't executing the generated code in the platform tests.  Did you want to take a stab at that?  That could be added here: https://github.com/flutter/packages/blob/master/packages/pigeon/platform_tests/flutter_null_safe_unit_tests/test/null_safe_test.dart

**Review by gaaclarke — COMMENTED** [review]
LGTM, one one suggestion is adding a platform unit test.

File: packages/pigeon/test/dart_generator_test.dart [hunk 2]
```diff
@@ -401,6 +445,51 @@ void main() {
     expect(code, contains('EnumClass doSomething(EnumClass arg0);'));
   });
 
+  test('flutter non-nullable enum argument with enum class', () {
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/pigeon/test/dart_generator_test.dart:448]: Why? This is testing the type itself, not a value. A type can't be non-null (except in the sense of not being the special null type in languages where there is one, but that's not what this is testing), it can only be non-nullable.

File: packages/pigeon/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,8 @@
+## 3.0.2
+
+* Fixes non-nullable classes and enums as fields.
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/pigeon/CHANGELOG.md:3]: Same; non-nullable is accurate.

(See also https://dart.dev/null-safety)

File: packages/pigeon/lib/dart_generator.dart [hunk 4]
```diff
@@ -192,7 +192,8 @@ final BinaryMessenger? _binaryMessenger;
         final String returnType = _makeGenericTypeArguments(func.returnType);
         final String castCall = _makeGenericCastCall(func.returnType);
         const String accessor = 'replyMap[\'${Keys.result}\']';
-        final String unwrapper = func.returnType.isNullable ? '' : '!';
+        final String unwrapper =
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/pigeon/lib/dart_generator.dart]: Renamed to `nullHandler`, which is the best name I could come up with for what this is doing.

File: packages/pigeon/pigeons/nullable_returns.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/pigeon/pigeons/nullable_returns.dart]: Added tests. I noticed that there were also no platform tests of the existing non-collection null return handling, so I added that too.

This spawned a few other changes:
- Renamed the null-return-type API classes in the Pigeon file, because they made no sense. Originally I had just copied the non-collection versions and hadn't paid close attention to the names, but when using them in the tests they were actively misleading, and it seemed pretty clear once I looked at them again that they were copypasta'd from another `pigeons/` file that was testing something completely different.
- Updated `build_runner` since I couldn't run the current version (to generate new mocks).

I'm now including the updated generated files, since otherwise the mocks and mock annotations that are now in this PR would refer to things that weren't in the versions of the files checked into the tree, which would be really weird.

File: packages/pigeon/test/dart_generator_test.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaaclarke) [reply packages/pigeon/test/dart_generator_test.dart:448]: I didn't realize that dart uses `non-nullable`.  That's confusing since other languages (java, objc) use "nullable/nonnull".

**Review by gaaclarke — APPROVED** [review]
Thanks for adding the tests, lgtm.
